src/ring_theory/polynomial/homogeneous.lean

```diff
@@ -0,0 +1,160 @@
+/-
+Copyleft 2020 Johan Commelin. No rights reserved.
+Authors: Johan Commelin
+-/
+
+import data.mv_polynomial
+import data.fintype.card
+
+/-!
+# Homogeneous polynomials
+
+A multivariate polynomial `φ` is homogeneous of degree `n`
+if all monomials occuring in `φ` have degree `n`.
+-/
+
+open_locale big_operators
+
+namespace mv_polynomial
+variables {σ : Type*} {τ : Type*} {R : Type*} {S : Type*}
+
+/-- A multivariate polynomial `φ` is homogeneous of degree `n`
+if all monomials occuring in `φ` have degree `n`. -/
+def is_homogeneous [comm_semiring R] (φ : mv_polynomial σ R) (n : ℕ) :=
+∀ ⦃d⦄, coeff d φ ≠ 0 → ∑ i in d.support, d i = n
+
+section
+variables [comm_semiring R]
+
+variables {σ R}
+
+lemma is_homogeneous_monomial (d : σ →₀ ℕ) (r : R) (n : ℕ) (hn : ∑ i in d.support, d i = n) :
+  is_homogeneous (monomial d r) n :=
+begin
+  intros c hc,
+  rw coeff_monomial at hc,
+  split_ifs at hc with h,
+  { subst c, exact hn },
+  { contradiction }
+end
+variables (σ) {R}
+
+lemma is_homogeneous_C (r : R) :
+  is_homogeneous (C r : mv_polynomial σ R) 0 :=
+begin
+  apply is_homogeneous_monomial,
+  simp only [finsupp.zero_apply, finset.sum_const_zero],
+end
+
+variables (σ R)
+
+lemma is_homogeneous_zero (n : ℕ) : is_homogeneous (0 : mv_polynomial σ R) n :=
+λ d hd, false.elim (hd $ coeff_zero _)
+
+lemma is_homogeneous_one : is_homogeneous (1 : mv_polynomial σ R) 0 :=
+is_homogeneous_C _ _
+
+variables {σ} (R)
+
+lemma is_homogeneous_X (i : σ) :
+  is_homogeneous (X i : mv_polynomial σ R) 1 :=
+begin
+  apply is_homogeneous_monomial,
+  simp only [finsupp.support_single_ne_zero one_ne_zero, finset.sum_singleton],
+  exact finsupp.single_eq_same
+end
+
+end
+
+namespace is_homogeneous
+variables [comm_semiring R] {φ ψ : mv_polynomial σ R} {m n : ℕ}
+
+lemma coeff_eq_zero (hφ : is_homogeneous φ n) (d : σ →₀ ℕ) (hd : ∑ i in d.support, d i ≠ n) :
+  coeff d φ = 0 :=
+by { have aux := mt (@hφ d) hd, classical, rwa not_not at aux }
+
+lemma inj_right (hm : is_homogeneous φ m) (hn : is_homogeneous φ n) (hφ : φ ≠ 0) :
+  m = n :=
+begin
+  obtain ⟨d, hd⟩ : ∃ d, coeff d φ ≠ 0 := φ.exists_coeff_ne_zero hφ,
+  rw [← hm hd, ← hn hd]
+end
+
+lemma add (hφ : is_homogeneous φ n) (hψ : is_homogeneous ψ n) :
+  is_homogeneous (φ + ψ) n :=
+begin
+  intros c hc,
+  rw coeff_add at hc,
+  obtain h|h : coeff c φ ≠ 0 ∨ coeff c ψ ≠ 0,
+  { contrapose! hc, simp only [hc, add_zero] },
+  { exact hφ h },
+  { exact hψ h }
+end
+
+lemma sum {ι : Type*} (s : finset ι) (φ : ι → mv_polynomial σ R) (n : ℕ)
+  (h : ∀ i ∈ s, is_homogeneous (φ i) n) :
+  is_homogeneous (∑ i in s, φ i) n :=
+begin
+  classical,
+  revert h,
+  apply finset.induction_on s,
+  { intro, simp only [is_homogeneous_zero, finset.sum_empty] },
+  { intros i s his IH h,
+    simp only [his, finset.sum_insert, not_false_iff],
+    apply (h i (finset.mem_insert_self _ _)).add (IH _),
+    intros j hjs,
+    exact h j (finset.mem_insert_of_mem hjs) }
+end
+
```

Comment:
Prove homogeneous polynomials of fixed degree form a submodule? Hopefully it's easy from here. Or it can wait.

Reply:
Hmm, may I should just make it a submodule to start with. But then I feel like I get sucked into defining graded structures on rings... and that seems like a separate project.